src/invariant/ConservationOfLumensTests.cpp

```diff
@@ -0,0 +1,321 @@
+            });
+}
+
+int64_t
+getCoinsAboveReserve(std::vector<LedgerEntry> const& entries,
+                     Application& app)
+{
+    return std::accumulate(
+            entries.begin(), entries.end(), static_cast<int64_t>(0),
+            [&app] (int64_t lhs, LedgerEntry const& rhs) {
+                auto& lm = app.getLedgerManager();
+                auto& account = rhs.data.account();
+                return lhs + account.balance -
+                       lm.getMinBalance(account.numSubEntries);
+            });
+}
+
+std::vector<LedgerEntry>
+updateBalances(std::vector<LedgerEntry> entries,
+               Application& app,
+               std::default_random_engine& gen,
+               int64_t netChange)
+{
+    int64_t initialCoins = getTotalBalance(entries);
+    int64_t pool = netChange + getCoinsAboveReserve(entries, app);
+    int64_t totalDelta = 0;
+    for (auto iter = entries.begin(); iter != entries.end(); ++iter)
+    {
+        auto& account = iter->data.account();
+        auto minBalance =
+            app.getLedgerManager().getMinBalance(account.numSubEntries);
+        pool -= account.balance - minBalance;
+
+        int64_t delta = 0;
+        if (iter+1 != entries.end())
+        {
+            int64_t maxDecrease = minBalance - account.balance;
+            int64_t maxIncrease = pool;
+            REQUIRE(maxIncrease >= maxDecrease);
+            std::uniform_int_distribution<int64_t> dist(maxDecrease, maxIncrease);
+            delta = dist(gen);
+        }
+        else
+        {
+            delta = netChange - totalDelta;
+        }
+
+        pool -= delta;
+        account.balance += delta;
+        totalDelta += delta;
+        REQUIRE(account.balance >= minBalance);
+    }
+    REQUIRE(totalDelta == netChange);
+
+    auto finalCoins = getTotalBalance(entries);
+    REQUIRE(initialCoins + netChange == finalCoins);
+
+    auto& lh = app.getLedgerManager().getCurrentLedgerHeader();
+    lh.totalCoins += netChange;
+    return entries;
+}
+
+std::vector<LedgerEntry>
+updateBalances(std::vector<LedgerEntry> const& entries,
+               Application& app,
+               std::default_random_engine& gen)
+{
+    int64_t coinsAboveReserve = getCoinsAboveReserve(entries, app);
+    auto& lh = app.getLedgerManager().getCurrentLedgerHeader();
+    std::uniform_int_distribution<int64_t> dist(
+            lh.totalCoins - coinsAboveReserve, INT64_MAX);
+    int64_t newTotalCoins = dist(gen);
+    return updateBalances(entries, app, gen, newTotalCoins - lh.totalCoins);
+}
+
+std::vector<EntryFrame::pointer>
+generateEntryFrames(std::vector<LedgerEntry> const& entries)
+{
+    std::vector<EntryFrame::pointer> result;
+    std::transform(entries.begin(), entries.end(), std::back_inserter(result),
+        [] (LedgerEntry const& le) {
+            return EntryFrame::FromXDR(le);
+        });
+    return result;
+}
+
+UpdateList
+generateUpdateList(std::vector<EntryFrame::pointer> const& current,
```

Comment:
same comment than earlier: I am not sure going the random route for this particular class of code is the most elegant way to go about it. Basically this code even randomized doesn't really tell us if the invariant will work all the time as the type of changes it's doing are always of the same kind (and it's the same than a simple fixed test)